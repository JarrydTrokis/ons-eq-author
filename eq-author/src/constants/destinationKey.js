export const NextPage = "NextPage";
export const Default = "Default";
export const EndOfCurrentSection = "EndOfCurrentSection";

export const destinationKey = {
  [EndOfCurrentSection]: "End of current section",
  [NextPage]: "Next page",
  [Default]: "Select a destination",
};
