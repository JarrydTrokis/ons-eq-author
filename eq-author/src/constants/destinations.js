export const NextPage = "NextPage";
export const Default = "Default";
export const EndOfCurrentSection = "EndOfCurrentSection";

export const logicalDestinations = () => [
  {
    id: NextPage,
    logicalDestination: NextPage,
    displayName: NextPage,
  },
  {
    id: EndOfCurrentSection,
    logicalDestination: EndOfCurrentSection,
    displayName: EndOfCurrentSection,
  },
];
