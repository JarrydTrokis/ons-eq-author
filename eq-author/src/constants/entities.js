export const SECTION = "section";
export const PAGE = "page";
export const QUESTION_CONFIRMATION = "question-confirmation";
export const INTRODUCTION = "introduction";
