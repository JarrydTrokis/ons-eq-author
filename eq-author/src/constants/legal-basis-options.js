const LEGAL_BASIS_OPTIONS = [
  {
    value: "NOTICE_1",
    description: "Section 1 of the Statistics of Trade Act 1947.",
  },
  {
    value: "NOTICE_2",
    description: "Sections 2 and 3 of the Statistics of Trade Act 1947.",
  },
  {
    value: "NOTICE_3",
    description: "Sections 3 and 4 of the Statistics of Trade Act 1947.",
  },
  {
    value: "NOTICE_NI",
    description:
      "Article 5 of the Statistics of Trade and Employment (Northern Ireland) Order 1988.",
  },
  {
    value: "NOTICE_FUELS",
    description:
      "Petroleum Act 1998 and Section 1 of the Statistics of Trade Act 1947.",
  },
  {
    value: "VOLUNTARY",
    description: "Voluntary",
  },
];

export default LEGAL_BASIS_OPTIONS;
