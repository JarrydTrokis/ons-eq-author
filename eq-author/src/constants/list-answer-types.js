export const DRIVING = "driving";
export const ANOTHER = "another";
