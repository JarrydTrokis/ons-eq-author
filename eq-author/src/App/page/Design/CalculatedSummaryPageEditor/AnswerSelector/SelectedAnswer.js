import React, { useState } from "react";
import PropType from "prop-types";
import styled from "styled-components";

import { colors, focusStyle } from "constants/theme";

import Truncated from "components/Truncated";
import { MenuItemType } from "components/ContentPickerv3/Menu";
import Modal from "components-themed/Modal";

import { ReactComponent as IconClose } from "assets/icon-close.svg";
import {
  DELETE_BUTTON_TEXT,
  DELETE_CALC_SUM_ANSWER_TITLE,
} from "constants/modal-content";

const Wrapper = styled.div`
  background: ${colors.primary};
  color: ${colors.white};
  border-radius: 3px;
  display: flex;
  align-items: center;
  justify-content: center;
  text-align: left;
  margin-bottom: 0.5em;
`;

const Title = styled(Truncated)`
  padding: 0.3em 0 0.3em 0.5em;
  line-height: 1.3;
`;

const Chip = styled(MenuItemType)`
  color: ${colors.text};
  float: right;
  font-size: 0.8em;
`;

const CloseButton = styled.button`
  background: transparent;
  padding: 0;
  margin: 0;
  border: none;
  cursor: pointer;

  svg {
    width: 2em;
    height: 2em;
    vertical-align: middle;
  }

  svg {
    g,
    path,
    polygon {
      fill: ${colors.white};
    }
  }

  &:hover {
    svg {
      g,
      path,
      polygon {
        fill: ${colors.highlightBlue};
      }
    }
  }

  &:focus {
    ${focusStyle}
  }
`;

const SelectedAnswer = ({
  displayName,
  properties,
  type: answerType,
  folderType,
  onRemove,
}) => {
  const unitType = properties.unit || false;

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  return (
    <>
      <Modal
        title={DELETE_CALC_SUM_ANSWER_TITLE}
        positiveButtonText={DELETE_BUTTON_TEXT}
        isOpen={showDeleteModal}
        onConfirm={onRemove}
        onClose={() => setShowDeleteModal(false)}
      />
      <Wrapper>
        <Title>{displayName}</Title>
        {unitType && <Chip data-test="unit-type">{unitType}</Chip>}
        {folderType === "ListCollectorFolder" && (
          <Chip>List collector follow-up</Chip>
        )}
        <Chip>{answerType}</Chip>
        <CloseButton
          data-test="remove-answer-button"
          onClick={() => setShowDeleteModal(true)}
        >
          <IconClose />
        </CloseButton>
      </Wrapper>
    </>
  );
};

SelectedAnswer.propTypes = {
  onRemove: PropType.func.isRequired,
  displayName: PropType.string.isRequired,
  type: PropType.string.isRequired,
  folderType: PropType.string,
  properties: PropType.object.isRequired, // eslint-disable-line
};

export default SelectedAnswer;
