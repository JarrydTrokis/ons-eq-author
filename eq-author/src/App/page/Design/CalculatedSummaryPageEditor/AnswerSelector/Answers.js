import React, { useState } from "react";
import styled from "styled-components";
import PropTypes from "prop-types";

import { colors } from "constants/theme";
import { calculatedSummaryErrors } from "constants/validationMessages";

import SelectedAnswer from "./SelectedAnswer";
import TextButton from "components/buttons/TextButton";
import Button from "components/buttons/Button";
import ValidationError from "components/ValidationError";
import Modal from "components-themed/Modal/modal";
import {
  DELETE_ALL_CALC_SUM_ANSWERS_TITLE,
  DELETE_BUTTON_TEXT,
} from "constants/modal-content";

import { getPageByAnswerId } from "utils/questionnaireUtils";
import { useQuestionnaire } from "components/QuestionnaireContext";

const Title = styled.h3`
  font-weight: bold;
  font-size: 0.9em;
  color: ${colors.darkGrey};
  margin: 0;
`;

const RemoveAllBtn = styled(TextButton)`
  letter-spacing: 0.05rem;
  font-weight: bold;
  font-size: 0.8rem;
  margin: 0 0 0 auto;
`;

const SelectButton = styled(Button)`
  width: 100%;
`;

const Header = styled.div`
  display: flex;
  margin-bottom: 1em;
`;

const Body = styled.div`
  margin-bottom: 1em;
`;

const ErrorWrapper = styled.div`
  ${(props) =>
    props.hasError &&
    `
      border: 2px solid ${colors.errorPrimary};
      padding: 1em;
    `}

  div:last-of-type {
    margin-bottom: 0;
  }
`;

const Footer = styled.div``;

// Function exported for testing
export const findAnswersErrors = (errors) => {
  return (
    errors.some((error) => error.errorCode === "ERR_NO_ANSWERS") ||
    errors.some(
      (error) => error.errorCode === "ERR_CALCULATED_UNIT_INCONSISTENCY"
    ) ||
    errors.some((error) => error.errorCode === "CALCSUM_MOVED")
  );
};

const Answers = ({ page, onUpdateCalculatedSummaryPage, onSelect }) => {
  const sectionTitle = page.section.displayName;
  const answerType = page.summaryAnswers[0].type;
  const selectedAnswers = page.summaryAnswers;

  const { questionnaire } = useQuestionnaire();

  const {
    validationErrorInfo: { errors },
  } = page;

  const [showDeleteModal, setShowDeleteModal] = useState(false);

  const selectedAnswersLeft = (answers) => {
    return selectedAnswers.filter(
      ({ id }) => !answers.find((answer) => answer.id === id)
    );
  };
  const handleRemoveAnswers = (answers) =>
    onUpdateCalculatedSummaryPage({
      id: page.id,
      summaryAnswers: selectedAnswersLeft(answers),
      type: selectedAnswersLeft(answers).length === 0 ? "" : answerType,
    });

  return (
    <>
      <Modal
        title={DELETE_ALL_CALC_SUM_ANSWERS_TITLE}
        positiveButtonText={DELETE_BUTTON_TEXT}
        isOpen={showDeleteModal}
        onConfirm={() => handleRemoveAnswers(selectedAnswers)}
        onClose={() => setShowDeleteModal(false)}
      />
      <Header>
        <Title>
          {answerType} answers in {sectionTitle}
        </Title>
        <RemoveAllBtn onClick={() => setShowDeleteModal(true)}>
          Remove all
        </RemoveAllBtn>
      </Header>
      <Body>
        <ErrorWrapper hasError={errors.length > 0 && findAnswersErrors(errors)}>
          {selectedAnswers.map((answer) => (
            <SelectedAnswer
              key={answer.id}
              insideListCollectorFolder={
                getPageByAnswerId(questionnaire, answer.id)?.folder?.listId !=
                null
              }
              {...answer}
              onRemove={() => handleRemoveAnswers([answer])}
            />
          ))}
        </ErrorWrapper>
        {errors.length > 0 && findAnswersErrors(errors) && (
          <ValidationError>
            {calculatedSummaryErrors[errors[0].errorCode]}
          </ValidationError>
        )}
      </Body>
      <Footer>
        <SelectButton variant="secondary" onClick={onSelect}>
          Select another {answerType.toLowerCase() || ""} answer or calculated
          summary total
        </SelectButton>
      </Footer>
    </>
  );
};

Answers.propTypes = {
  page: PropTypes.object.isRequired, // eslint-disable-line
  onUpdateCalculatedSummaryPage: PropTypes.func.isRequired,
  onSelect: PropTypes.func.isRequired, // eslint-disable-line
};

export default Answers;
