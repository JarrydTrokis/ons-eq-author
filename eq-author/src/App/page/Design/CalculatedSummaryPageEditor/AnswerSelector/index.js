import React, { useMemo, useState } from "react";
import gql from "graphql-tag";
import styled from "styled-components";
import PropType from "prop-types";

import { colors } from "constants/theme";
import { NUMBER, CURRENCY, UNIT, PERCENTAGE } from "constants/answer-types";

import Answers from "./Answers";
import Empty from "./Empty";
import AnswerPicker from "components/AnswerPicker";

import { useQuestionnaire } from "components/QuestionnaireContext";
import getContentBeforeEntity from "utils/getContentBeforeEntity";

const Container = styled.div`
  border: 1px solid ${colors.grey};
  border-radius: 3px;
  padding: 1em;
  margin-bottom: 2em;
`;

export const filterAvailableAnswers = (entity) =>
  [CURRENCY, UNIT, PERCENTAGE, NUMBER].includes(entity.type) ? entity : [];

const insertSection = (sections, section, folder, page, answers) => {
  sections.push({
    ...section,
    folders: [],
  });
  sections = insertFolder(sections, folder, page, answers);
  return sections;
};

const insertFolder = (sections, folder, page, answers) => {
  sections[sections.length - 1].folders.push({
    ...folder,
    pages: [],
  });
  sections = insertPage(sections, page, answers);
  return sections;
};

const insertPage = (sections, page, answers) => {
  sections[sections.length - 1].folders[
    sections[sections.length - 1].folders.length - 1
  ].pages.push({
    ...page,
    answers,
  });
  return sections;
};

const AnswerSelector = ({ page, onUpdateCalculatedSummaryPage }) => {
  const [showPicker, setShowPicker] = useState(false);

  const { questionnaire } = useQuestionnaire();

  const getAllCalculatedSummaryPages = (pageId, sectionId) => {
    const sections = [];

    for (const section of questionnaire.sections) {
      if (section.id === sectionId) {
        return sections;
      }
      for (const folder of section.folders) {
        for (const page of folder.pages) {
          const answers = page?.answers;
          if (page.pageType === "CalculatedSummaryPage") {
            if (page.id === pageId) {
              return sections;
            }
            if (sections[sections.length - 1]?.id === section?.id) {
              if (
                sections[sections.length - 1].folders[
                  sections[sections.length - 1].folders.length - 1
                ]?.id === folder?.id
              ) {
                insertPage(sections, page, answers);
              } else {
                insertFolder(sections, folder, page, answers);
              }
            } else {
              insertSection(sections, section, folder, page, answers);
            }
          }
        }
      }
    }
    return sections;
  };

  const calculatedSummaries = getAllCalculatedSummaryPages(
    page.id,
    page.section.id
  );

  const availableAnswers =
    (
      questionnaire &&
      getContentBeforeEntity({
        questionnaire,
        id: page.id,
        preprocessAnswers: filterAvailableAnswers,
      })
    )?.filter((section) => section.id === page.section.id) || [];

  const calculatedSummaryAnswers = [
    ...calculatedSummaries,
    ...availableAnswers,
  ];

  const availableSummaryAnswers = useMemo(
    () => calculatedSummaryAnswers,
    [questionnaire, page.id, page.section.id]
  );

  const handlePickerOpen = () => setShowPicker(true);
  const handlePickerClose = () => setShowPicker(false);
  const handlePickerSubmit = (answers) => {
    handlePickerClose();
    onUpdateCalculatedSummaryPage({
      id: page.id,
      summaryAnswers: answers,
      type: answers[0].type,
    });
  };

  return (
    <>
      <Container>
        {page.summaryAnswers.length > 0 ? (
          <Answers
            page={page}
            onUpdateCalculatedSummaryPage={onUpdateCalculatedSummaryPage}
            onSelect={handlePickerOpen}
          />
        ) : (
          <Empty
            page={page}
            availableSummaryAnswers={availableSummaryAnswers}
            onSelect={handlePickerOpen}
          />
        )}
      </Container>
      <AnswerPicker
        isOpen={showPicker}
        onClose={handlePickerClose}
        onSubmit={handlePickerSubmit}
        startingSelectedAnswers={page.summaryAnswers}
        data={availableSummaryAnswers}
        questionnaire={questionnaire}
        title="Select an answer or calculated summary total"
        showTypes
      />
    </>
  );
};

AnswerSelector.fragments = {
  AnswerSelector: gql`
    fragment AnswerSelector on CalculatedSummaryPage {
      id
      section {
        id
        displayName
      }
      summaryAnswers {
        id
        displayName
        type
        properties
      }
      type
    }
  `,
};

AnswerSelector.propTypes = {
  page: PropType.object.isRequired, // eslint-disable-line
  onUpdateCalculatedSummaryPage: PropType.func.isRequired,
};

export default AnswerSelector;
