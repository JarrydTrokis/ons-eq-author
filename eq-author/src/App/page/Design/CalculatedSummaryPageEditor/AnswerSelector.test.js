import React from "react";
import { shallow } from "enzyme";

import { NUMBER } from "constants/answer-types";
import ContentPickerModal from "components/ContentPickerModal";

import AnswerSelector, { SuggestionButton } from "./AnswerSelector";
import AnswerChip from "./AnswerChip";

describe("AnswerSelector", () => {
  let mockHandlers, page, answers;
  beforeEach(() => {
    mockHandlers = {
      fetchAnswers: jest.fn(),
      onChange: jest.fn(),
      onUpdate: jest.fn(),
      onChangeUpdate: jest.fn(),
      onUpdateCalculatedSummaryPage: jest.fn(),
    };
    page = {
      id: "2",
      title: "",
      alias: "",
      pageType: "CalculatedSummaryPage",
      position: 2,
      displayName: "Foo",
      totalTitle: "",
      section: { id: "1", displayName: "This Section" },
      summaryAnswers: [],
      availableSummaryAnswers: [],
    };
    answers = [
      {
        id: 1,
        displayName: "Answer 1",
        type: NUMBER,
        page: {
          id: "1",
          displayName: "Foo",
          section: { id: 1, displayName: "Section 1" },
        },
      },
      {
        id: 2,
        displayName: "Answer 2",
        type: NUMBER,
        page: {
          id: "1",
          displayName: "Foo",
          section: { id: 1, displayName: "Section 1" },
        },
      },
      {
        id: 3,
        displayName: "Answer 3",
        type: NUMBER,
        page: {
          id: "1",
          displayName: "Foo",
          section: { id: 1, displayName: "Section 1" },
        },
      },
    ];
  });

  it("should render", () => {
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should render a list of AnswerChips when answers selected", () => {
    page.summaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(AnswerChip)).toHaveLength(3);
  });

  it("can add a single answer", () => {
    page.availableSummaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    wrapper.find(ContentPickerModal).simulate("submit", { value: { id: 1 } });
    expect(mockHandlers.onUpdateCalculatedSummaryPage).toHaveBeenCalledWith({
      id: "2",
      summaryAnswers: [{ value: { id: 1 } }],
    });
  });

  it("should remove single answers on remove click", () => {
    page.summaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(AnswerChip)).toHaveLength(3);
    wrapper
      .find(AnswerChip)
      .first()
      .simulate("remove");
    expect(mockHandlers.onUpdateCalculatedSummaryPage).toHaveBeenCalledWith({
      id: "2",
      summaryAnswers: [answers[1], answers[2]],
    });
  });

  it("should remove multiple answers on removeAll click", () => {
    page.summaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(AnswerChip)).toHaveLength(3);
    wrapper.find('[data-test="remove-all"]').simulate("click");
    expect(mockHandlers.onUpdateCalculatedSummaryPage).toHaveBeenCalledWith({
      id: "2",
      summaryAnswers: [],
    });
  });

  it("should add many answers when a suggestion is selected", () => {
    page.availableSummaryAnswers = answers;
    const wrapper = shallow(<AnswerSelector page={page} {...mockHandlers} />);
    expect(wrapper.find(SuggestionButton)).toHaveLength(1);
    wrapper.find(SuggestionButton).simulate("click");
    expect(mockHandlers.onUpdateCalculatedSummaryPage).toHaveBeenCalledWith({
      id: "2",
      summaryAnswers: [
        expect.objectContaining({
          displayName: "Answer 1",
          id: 1,
          type: NUMBER,
        }),
        expect.objectContaining({
          displayName: "Answer 2",
          id: 2,
          type: NUMBER,
        }),
        expect.objectContaining({
          displayName: "Answer 3",
          id: 3,
          type: NUMBER,
        }),
      ],
    });
  });
});
