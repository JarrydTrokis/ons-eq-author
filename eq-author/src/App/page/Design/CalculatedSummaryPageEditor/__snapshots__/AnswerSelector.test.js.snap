// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerSelector should render when there are available summary answers 1`] = `
<div>
  <AnswerSelector__Box>
    <AnswerSelector__Answers>
      <div>
        <AnswerSelector__Empty>
          <AnswerSelector__EmptyTitle>
            No answers selected
          </AnswerSelector__EmptyTitle>
          <AnswerSelector__EmptyText>
            Select an answer using the button below.
          </AnswerSelector__EmptyText>
          <AnswerSelector__EmptyButton
            data-test="answer-selector-empty"
            disabled={false}
            onClick={[Function]}
            small={true}
          >
            Select an answer
          </AnswerSelector__EmptyButton>
        </AnswerSelector__Empty>
      </div>
      <CalSumContentPicker
        data={
          Array [
            Object {
              "displayName": "Section 1",
              "id": 1,
              "pages": Array [
                Object {
                  "answers": Array [
                    Object {
                      "displayName": "Answer 1",
                      "id": 1,
                      "type": "Number",
                    },
                    Object {
                      "displayName": "Answer 2",
                      "id": 2,
                      "type": "Number",
                    },
                    Object {
                      "displayName": "Answer 3",
                      "id": 3,
                      "type": "Number",
                    },
                  ],
                  "displayName": "Foo",
                  "id": "1",
                },
              ],
            },
          ]
        }
        isOpen={false}
        onClose={[Function]}
        onSubmit={[Function]}
        startingSelectedAnswers={Array []}
      />
    </AnswerSelector__Answers>
  </AnswerSelector__Box>
</div>
`;

exports[`AnswerSelector should render when there are no available summary answers 1`] = `
<div>
  <AnswerSelector__Box>
    <AnswerSelector__Answers>
      <div>
        <AnswerSelector__Empty>
          <AnswerSelector__EmptyTitle>
            No answers available
          </AnswerSelector__EmptyTitle>
          <AnswerSelector__EmptyText>
            There are no answers to provide a calculated summary.
          </AnswerSelector__EmptyText>
          <AnswerSelector__EmptyButton
            data-test="answer-selector-empty"
            disabled={true}
            onClick={[Function]}
            small={true}
          >
            Select an answer
          </AnswerSelector__EmptyButton>
        </AnswerSelector__Empty>
      </div>
      <CalSumContentPicker
        data={Array []}
        isOpen={false}
        onClose={[Function]}
        onSubmit={[Function]}
        startingSelectedAnswers={Array []}
      />
    </AnswerSelector__Answers>
  </AnswerSelector__Box>
</div>
`;
